cpp/onedal/detail/array_impl.hpp

```diff
@@ -16,43 +16,39 @@
 
 #pragma once
 
-#include "onedal/common.hpp"
-#include "onedal/detail/common.hpp"
-#include "onedal/detail/table_data.hpp"
+#include "onedal/detail/deleters.hpp"
+#include "onedal/detail/table_impl.hpp"
 
 namespace dal {
 namespace detail {
 
-// TODO: this is particular array impl for slices,
-// rename it
 template <typename T>
 class array_impl {
 public:
-    array_impl(const table_data_ptr& data_origin, const range2d& slice)
-        : _data_origin(data_origin),
-          _slice(slice) {
-        _data = _data_origin->get_data_ptr(_slice, _data);
-    }
+    template <typename Deleter>
+    array_impl(T* data, std::int64_t size, Deleter d)
+        : _data(data),
+          _size(size),
+          _deleter(shared<deleter_iface>(new deleter_container<Deleter>(d)))
+    { }
 
     ~array_impl() {
-        _data_origin->release_data_ptr(_slice, _data, true);
+        _deleter->operator()((void*)_data);
```

Comment:
```suggestion
        (*_deleter)(_data);
        _deleter->operator()((void*)_data);
```
I suggest to not cast data to void explicitly, since deleter may know the pointer type or be template itself.